### ISSUE TITLE
Add 'events' to widgets

### ISSUE THREAD

**Issue by martinohanlon** [lawsie/guizero#120]
There are a couple of open PR's relating to adding events to widgets. #14 and #118 .

Its clearly a good idea, so I have opened this issue to keep track.

A couple of initial points / questions come to mine

1. The implementation of this should be via a mixin, its a shared function of many widgets

2. There is a choice re interface, functions or properties. 

tk currently uses functions like `on_mouse_click` to set a command, this is familiar to tk users.

other "zero" libraries have moved to using properties to set a command, e.g. gpiozero

```
def my_func():
    print("hi")
button = Button(17)
button.when_pressed = my_func
```

If implemented as properties (which I think is a good idea) should they be `when_mouse_clicked`, `when_key_pressed`, `when_mouse_moved` - its a bit more verbose, but it is simple to understand.

This also lends its self to implementing other events such as `when_pressed` for a PushButton or `when_changed` for a TextBox.

3. The tk event callbacks will return `Event` objects [http://effbot.org/tkinterbook/tkinter-events-and-bindings.htm](http://effbot.org/tkinterbook/tkinter-events-and-bindings.htm) - is this the right thing to do for guizero? 

One of the problems is that the `Event` object is a tk object not a guizero object, so (as an example) it refers to the tk widget not the guizero widget.

I think the right approach is to create an Event structure for guizero which returns guizero widgets but could also be extended to include other useful values in the future e.g. "imagine a `when_changed` event on a `Slider`" would return the Event structure but also the value the Slider changed too.

**Comment by Coal0** [comment]
Not a huge fan of `when_<event>`, I think `on_<event>` simply sounds better. This is a trivial issue, though.

A custom `Event` object seems like a great idea. I envision something like this:

* `MouseEvent`: pointer (x, y) coordinates (`tk.Event.x`, `tk.Event.y`). Should probably only be returned on an actual mouse 'click' event. If there's a need for other mouse-related events, like scrolling, maybe add `ScrollEvent`?

* `KeyboardEvent`: key pressed (`tk.Event.char`, not `tk.Event.keysym`). Returned on a keypress event.

Begs the question, how many events should we implement? 99% of use cases are covered by:

* Left mouse button
* Right mouse button
* Keypress

---

```python
class KeyboardEvent:
    def __init__(self, event):
        self._key = event.char

    @property
    def key(self):
        return self._key


class MouseEvent:
    def __init__(self, event):
       self._x, self._y = event.x, event.y

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y
```

Psuedo implementation:

```python
class Widget:
    ...

    @property
    def on_left_mouse_click(self):
       # on_primary_mouse_click?
        return self._event_left_mouse_click

    @on_left_mouse_click.setter
    def on_left_mouse_click(self, callable):
        self.tk.bind("<Button-1>", callable)
        self._event_left_mouse_click = callable

    @property
    def on_right_mouse_click(self):
       # on_secondary_mouse_click?
        return self._event_right_mouse_click

    @on_right_mouse_click.setter
    def on_right_mouse_click(self, callable):
        self.tk.bind("<Button-2>", callable)
        self._event_right_mouse_click = callable

>>> widget = Widget()
>>> widget.on_left_mouse_click = lambda event: print(event.x, event.y)
>>> # Left mouse click
XXX, YYY

```

---

Another concern: Do we use `self.tk.winfo_toplevel().bind` or simply `self.tk.bind`?
Let's save everyone a headache and do this right. For some bizarre reason, it's sometimes required to bind to the toplevel widget.

**Comment by lawsie** [comment]
I prefer "when" - it 's more easily understandable for kids and fits in with statements like "when green flag pressed" which they may have encountered before in things like Scratch. I also think that the names of the events are extremely important so that kids understand what they are.

@drussell1974 did a lot of work on this in #14 and I'm sorry that I never got around to adding it :( 

I'm not sure re: the Event object but the simpler the thing that is returned the better, even if that means the functionality is reduced. It won't get used if kids can't understand how to unpack information from an object they aren't familiar with. 

Basically, let's keep kids in mind as the target audience! :)

**Comment by Coal0** [comment]
Okay, how about an (x, y) `tuple` and a key `str`?

**Comment by martinohanlon** [comment]
As a first implementation I am planning on doing the following events:

`when_clicked`, `when_mouse_over`, `when_changed`, `when_key_pressed`, `when_got_focus`

**Comment by lawsie** [comment]
I think that 'focus' (/blur) is not a thing kids understand (having experimented with this a bit in Javascript).

**Comment by martinohanlon** [comment]
It does seem a little ambiguous. I dont like blur either I prefer `lost_focus`.

However.... `focus` is consistent with the widget method to give it "focus" i.e. `widget.focus()`.

It is also a well understood term and is one of the consistencies that tkinter shares with other user interface libraries.

If guizero was going to break with this tradition I would suggest:
 - `widget.select()` 
 - `widget.when_selected`
 - `widget.when_deselected`

Im not sure about it though, are widgets always selected when they are given focus....

**Comment by lawsie** [comment]
Well I actually meant maybe it isn't necessary in guizero at all? I don't think many kids would use it because they don't understand what focus is, and anyone who wanted to use it could potentially just use it from tkinter? I guess what I'm trying to say is that often I've found teachers don't use things because there are **too many** choices and they are overwhelmed. Do we need it? What is its purpose?

**Comment by martinohanlon** [comment]
I see..  It turns out `focus` is not that easy to implement so its already slipped right down the priority list.. 

'when_focus` would be useful for creating gui's which direct the user, highlighting the component which needs to inputted.

**Comment by martinohanlon** [comment]
In v0.5.0 I have introduced the following events to all the widgets:

- `when_clicked`
- `when_left_button_pressed`
- `when_left_button_released`
- `when_right_button_pressed`
- `when_right_button_released`
- `when_key_pressed`
- `when_key_released`
- `when_mouse_enters`
- `when_mouse_leaves`
- `when_mouse_dragged`

They are assigned callbacks functions like this:

```python
def clicked():
    print("clicked")

widget.when_clicked = clicked
```

The function which is called can also accept a parameter and will be passed data about the event which occured.

The event data returned has:

- `widget` - the guizero widget which raised the event
- `tk_event` - the [tkinter event object](http://effbot.org/tkinterbook/tkinter-events-and-bindings.htm)
- `key` - the key which raised the event
- `x` - the mouse's x position relative to the widget when the event occured
- `y` - the mouse's y position relative to the widget when the event occured
- `display_x` - the mouse's x position on the display when the event occured
- `display_y` - the mouse's y position on the display when the event occured

```python
def clicked(event_data):
    print("widget clicked = " + event_data.widget)
    print("mouse position = " + event_data.x + "." + event_data.y)
    
widget.when_clicked = clicked
```

The implementation was pretty complicated, due to many reasons such as:

- guizero widgets already used some tk events and you cant bind 2 callbacks to 1 tk event
- some guizero widgets have more than 1 tk object which needed to bind to a single callback
- some guizero widgets the self.tk object wasnt the object which needed to be bound to events.

... so I have noted in the documentation that this functionality is experimental as it might need to be changed.

**Comment by martinohanlon** [comment]
resolved in v0.5.0

**Comment by MrAHolmes** [comment]
I am trying to code a combo box that when changed alters the colour of another text box.
The combo box widget really needs a when_changed event.
The existing events only track over the un-expanded COMBO box.
This is messy as you have to position your mouse OVER the original combo box for mouse events to register AFTER you have actually made a change.

`
`from guizero import App, Combo, TextBox,Text

lst_Colours = ['Gold',
               'Gray',
               'Black',
               'Cyan',
               'Magenta',
               'Snow',
               'Yellow',
               'Red',
               'Orange',
               'Salmon']

app = App(title='Combo Box Event Issue',width=500,layout = 'grid')
 

lbl_promptColour = Text(app,text="Colour",grid=[1,0],width=15)
cbo_Colours = Combo(app,options=lst_Colours,selected=lst_Colours[0],grid=[1,1],width=15)
txt_ColourSwatch = Text(app,text="Colour \nSwatch",grid=[1,2],
                        bg=cbo_Colours.value,width=15,height=2,enabled=False)

   
def updateColourSwatch():
    txt_ColourSwatch.bg = cbo_Colours.value
    if cbo_Colours.value == "Black":
        txt_ColourSwatch.text_color = "White"
    else:
        txt_ColourSwatch.text_color = "Black"
    print(txt_ColourSwatch.bg,txt_ColourSwatch.text_color)

cbo_Colours.when_left_button_released = updateColourSwatch
cbo_Colours.when_mouse_leaves = updateColourSwatch

app.display()



`

**Comment by martinohanlon** [comment]
Just used the standard `command` on the Combo box. You don't need to use all the custome events.

```
cbo_Colours = Combo(app,options=lst_Colours,selected=lst_Colours[0],grid=[1,1],width=15,command=updateColourSwatch)
```

^ I haven't tested this code, but that is what the Combo command does.
